sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -388,6 +388,26 @@ abstract class BinaryComparison extends BinaryOperator with Predicate {
       defineCodeGen(ctx, ev, (c1, c2) => s"${ctx.genComp(left.dataType, c1, c2)} $symbol 0")
     }
   }
+
+  // The type checking logic of BinaryComparison can't be expressed by `inputType`, so we use
+  // AnyDataType here to bypass this check, and implement the actual type checking logic in
+  // `checkInputDataTypes`
+  override def inputType: AbstractDataType = AnyDataType
+
+  // TODO: although map type is not orderable, technically map type should be able to be used
+  // in equality comparison, update this type check once we support it.
+  override def checkInputDataTypes(): TypeCheckResult = {
+    super.checkInputDataTypes() match {
+      case TypeCheckResult.TypeCheckSuccess =>
+        if (RowOrdering.isOrderable(left.dataType)) {
```

Comment:
let me minimize the size of this PR, I'll keep `GreaterThan`, `LessThan`, etc. unchanged, and only update `EqualTo`, `EqualNullSafe` to forbid map types.